host/continue_as_new_test.go

```diff
@@ -333,35 +332,40 @@ GetHistoryLoop:
 
 		firstEvent := h.Events[0]
 		lastEvent := h.Events[len(h.Events)-1]
-		if lastEvent.GetEventType() != enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT {
-			if lastEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW {
-				// Ensure that timeout is not caused by runTimeout
-				s.True(timestamp.TimeValue(lastEvent.EventTime).Sub(timestamp.TimeValue(firstEvent.EventTime)) < 5*time.Second)
-			}
 
-			// Only PollForWorkflowTask if the last event is WorkflowTaskScheduled
-			if lastEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
-				s.Logger.Info(fmt.Sprintf("Execution not timed out yet. PollForWorkflowTask.  Last event is %v", lastEvent))
-				_, err := poller.PollAndProcessWorkflowTaskWithoutRetry(true, false)
-				s.Logger.Info("PollAndProcessWorkflowTask", tag.Error(err))
-				// Excluding ErrWorkflowCompleted because workflow might timeout while we are processing wtHandler.
-				if err != nil && err != matching.ErrNoTasks && err.Error() != consts.ErrWorkflowCompleted.Error() {
-					s.NoError(err)
-				}
-			}
+		if lastEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT {
+			s.Logger.Info("Workflow execution timedout.  Printing history for last run:")
+			common.PrettyPrintHistory(h, s.Logger)
 
-			time.Sleep(200 * time.Millisecond)
-			continue GetHistoryLoop
+			workflowComplete = true
+			break
 		}
 
-		s.Logger.Info("Workflow execution timedout.  Printing history for last run:")
-		common.PrettyPrintHistory(h, s.Logger)
+		if lastEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW {
+			// Ensure that timeout is not caused by runTimeout
+			s.True(timestamp.TimeValue(lastEvent.EventTime).Sub(timestamp.TimeValue(firstEvent.EventTime)) < 5*time.Second)
+		}
 
-		workflowComplete = true
-		break GetHistoryLoop
+		// Only PollForWorkflowTask if the last event is WorkflowTaskScheduled and we have at least 2 seconds left
+		// (to account for potential delay from queueing and matching task forwarding).
+		expiration := firstEvent.GetWorkflowExecutionStartedEventAttributes().GetWorkflowExecutionExpirationTime()
+		if lastEvent.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED &&
+			expiration.Sub(time.Now()) > 2*time.Second {
```

Comment:
Instead of keep checking and skipping, could we just wait?
time.Sleep(2 * time.Second - expiration.Sub(time.Now()))

Reply:
Not a big difference, but sure